### PR TITLE
Restricting DatePicker to select date from outside the defined Range (Min/Max). Added support for fhir path expression in Min / Max Validators

### LINKED ISSUE #1040: Add ability to validate the min/max for Date on a FHIR Questionnaire
**Is your feature request related to a problem? Please describe.**
Add ability to limit the date input values to either the minimum or maximum date allowed for a given use case e.g. 
1. A task should not start before a certain date 
2. When adding DOB, the DOB should be less than the current day/today 
3. When scheduling you can validate when a given task should be done or assigned to a given practitioner 

**Describe the solution you'd like**
1. To be able to resolve expressions such as today() in the min-max value extension
2. To use the min-max value extension in the actual date picker pop-up. at the moment the pop-up dialogue allows the user to put in any date and the error message only kicked in once the user has input a date.

**Describe alternatives you've considered**
1.  It would be much better if the date picker already doesn't allow the user to input a date outside the allowed range.

**Additional context**
A detailed use case is defined on this ticket https://github.com/opensrp/fhircore/issues/845 on OpenSRP FHIR Core repo

**Would you like to work on the issue?**
@RaaziaTarique

### PR DESCRIPTION
**Pull request by aurangzaibumer**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1040 

**Description**
The min/max validators are updated to use fhir path expressions. For date picker specifically, the widget doesn't allow input outside of the date range, this is a widget specific behavior. User can specify minimum and maximum date range using minValue and maxValue expression in questionnaire in order to prevent selecting the dates from outside the specified range. cqf-calculatedvalue expression is used to make the expression dynamic and use for some purpose as Age case (for example we can only registered a child that is born after 2020) (min value case).
 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MaxValueConstraintValidator.kt [hunk 1]
```diff
@@ -29,10 +29,13 @@ internal object MaxValueConstraintValidator :
     predicate = {
       extension: Extension,
       answer: QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent ->
-      answer.value > extension.value
+      answer.value > getExtensionValue(extension)
```
[the remark this answers is not in the document]
Comment (aurangzaibumer) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MaxValueConstraintValidator.kt:32]: checked, this is inclusive. Also checked the Json by adding dates of today minus day & today plus day. 
http://hl7.org/fhir/R4/extension-minvalue.html

File: demo/src/main/assets/new-patient-registration-paginated.json [hunk 2]
```diff
@@ -125,6 +125,26 @@
                 "expression": "Patient.birthDate",
                 "name": "patientBirthDate"
               }
+            },
+            {
+              "url" : "http://hl7.org/fhir/StructureDefinition/minValue",
+              "valueDate" : "",
+              "_valueDate" : {
+                "extension" : [
+                  {
```
[the remark this answers is not in the document]
Comment (aurangzaibumer) [reply demo/src/main/assets/new-patient-registration-paginated.json:133]: `  "extension" : [
              {
                "url" : "http://hl7.org/fhir/StructureDefinition/minValue",
                "valueDecimal" : 1,
                "_valueDecimal" : {
                  "extension" : [
                    {
                      "url" : "http://hl7.org/fhir/StructureDefinition/cqf-calculatedValue",`

I checked this from here https://build.fhir.org/ig/HL7/sdc/examples.html#cqf - please correct me if i'm wrong. thanks

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt [hunk 3]
```diff
@@ -29,11 +32,29 @@ internal object MinValueConstraintValidator :
...
+    extension.value.extension.firstOrNull()?.let {
+      ResourceMapper.fhirPathEngine
+        .evaluate(extension, (it.value as Expression).expression)
+        .firstOrNull()
+        ?.let { result = it as Type }
+    }
+  } else {
+    result = extension.value
+  }
+  return result!!
```
[the remark this answers is not in the document]
Comment (aurangzaibumer) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt:57]: MoreQuestItemComp is returning exlusively a DateType, here we are returning as a Type so we don't have to case it on the DatePickerFactory as DateType because it needs a dateType. Please let me know if you have a better idea on this. thanks

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 4]
```diff
@@ -71,7 +71,7 @@ import org.hl7.fhir.r4.utils.StructureMapUtilities
  */
 object ResourceMapper {
 
-  private val fhirPathEngine: FHIRPathEngine =
+  val fhirPathEngine: FHIRPathEngine =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:74]: revert this change.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt [hunk 5]
```diff
@@ -29,11 +32,29 @@ internal object MinValueConstraintValidator :
...
+        getExtensionValue(extension).primitiveValue()
+      )
     }
   )
 
+internal fun getExtensionValue(extension: Extension): Type {
+  var result: Type? = null
+  if (extension.value.hasExtension()) {
+    extension.value.extension.firstOrNull()?.let {
+      ResourceMapper.fhirPathEngine
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt:49]: create your own fhir path engine in this file.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt [hunk 6]
```diff
@@ -96,4 +101,88 @@ class MinValueConstraintValidatorTest {
...
+            url = MIN_VALUE_EXTENSION_URL
+            this.setValue(
+              DateType().apply {
+                extension =
+                  listOf(
+                    Extension(
+                      CQF_CALCULATED_EXPRESSION_URL,
+                      Expression().apply {
+                        language = "text/fhirpath"
+                        expression = "today() - 1 'days'"
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt:120]: i think this would not be necessarily stable. is there a way for you to make this more stable?

can you set the system date in the test or something? try this: https://www.baeldung.com/java-override-system-time

File: demo/src/main/assets/new-patient-registration-paginated.json [hunk 7]
```diff
@@ -125,6 +125,36 @@
                 "expression": "Patient.birthDate",
                 "name": "patientBirthDate"
               }
+            },
```
Comment (jingtang10) [inline demo/src/main/assets/new-patient-registration-paginated.json:128]: revert this change.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt [hunk 8]
```diff
@@ -29,11 +32,29 @@ internal object MinValueConstraintValidator :
...
+        R.string.min_value_validation_error_msg,
+        getExtensionValue(extension).primitiveValue()
+      )
     }
   )
 
+internal fun getExtensionValue(extension: Extension): Type {
+  var result: Type? = null
+  if (extension.value.hasExtension()) {
+    extension.value.extension.firstOrNull()?.let {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt:48]: you need to check the extension has the url: CQF_CALCULATED_EXPRESSION_URL

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt [hunk 9]
```diff
@@ -29,11 +32,29 @@ internal object MinValueConstraintValidator :
...
     { extension: Extension, context: Context ->
-      context.getString(R.string.min_value_validation_error_msg, extension.value.primitiveValue())
+      context.getString(
+        R.string.min_value_validation_error_msg,
+        getExtensionValue(extension).primitiveValue()
+      )
     }
   )
 
+internal fun getExtensionValue(extension: Extension): Type {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt:45]: i think it might be better to move this out of this file. just create a util file in the same package.

Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidator.kt:45]: can you change this to an extension property?

```
internal val Type.valueOrCalculatedValue
```

and then when you need to use this you do:

```
extension.value.valueOrCalculatedValue
```

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  private lateinit var viewHolder: QuestionnaireItemViewHolder
+  @Before
+  fun setup() {
+    activityScenarioRule.scenario.onActivity { activity -> parent = FrameLayout(activity) }
+    viewHolder = QuestionnaireItemDatePickerViewHolderFactory.create(parent)
+    setTestLayout(viewHolder.itemView)
+  }
+
+  @Test
+  fun shouldSetDateInput() {
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:63]: can you add a test case where this would fail?

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    val questionnaireItemView =
+      QuestionnaireItemViewItem(
+        Questionnaire.QuestionnaireItemComponent().apply {
+          addExtension().apply {
+            url = "http://hl7.org/fhir/StructureDefinition/minValue"
+            setValue(DateType(2020, 0, 1))
+          }
+          addExtension().apply {
+            url = "http://hl7.org/fhir/StructureDefinition/maxValue"
+            setValue(DateType(2025, 0, 1))
```
Comment (jingtang10) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:73]: please change this to something stable.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt [hunk 12]
```diff
@@ -229,6 +237,31 @@ val Questionnaire.QuestionnaireItemComponent.enableWhenExpression: Expression?
...
+  } else {
+    it.value as DateType
+  }
+}
+
+fun Questionnaire.QuestionnaireItemComponent.maximumValueForDate(): DateType? {
+  return this.extension.firstOrNull { it.url == MAX_VALUE_EXTENSION_URL }?.let {
+    processCQLExtensionForDate(it)
+  }
+}
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:263]: can you make sure this logic is in the validators, and maybe open up some internal functions.

and then when you're using this in the view holder factory you can just do a cast.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 13]
```diff
@@ -113,8 +116,23 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
         return MaterialDatePicker.Builder.datePicker()
           .setTitleText(R.string.select_date)
           .setSelection(selectedDate)
+          .setCalendarConstraints(getCalenderConstraint())
           .build()
       }
+
+      private fun getCalenderConstraint(): CalendarConstraints {
+        val min = questionnaireItemViewItem.questionnaireItem.minimumValueForDate()?.value?.time
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:124]: this should look like this:

```
val min = MinValueConstraintValidator.getMinValue(questionnaireItemViewItem.questionnaireItem) as Date
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]
thanks!

**Comment by aurangzaibumer** [comment]
https://user-images.githubusercontent.com/35099184/180433662-5c6fde87-bbcf-4558-b2f7-ad1d47f7eeb7.mp4

I've resolved _non-test case classes_ feedback, can you please review so that i can move to resolve the unit test cases. thanks in advance. @jingtang10 

cc @f-odhiambo

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import ca.uhn.fhir.context.support.DefaultProfileValidationSupport
+import com.google.android.fhir.datacapture.CQF_CALCULATED_EXPRESSION_URL
+import org.hl7.fhir.r4.hapi.ctx.HapiWorkerContext
+import org.hl7.fhir.r4.model.Expression
+import org.hl7.fhir.r4.model.Type
+import org.hl7.fhir.r4.utils.FHIRPathEngine
+
+object ValidationUtil {
+
+  internal val fhirPathEngine: FHIRPathEngine =
```
Comment (shoaibmushtaq25) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt:30]: Make this `private` as this is only used in this file.
```suggestion
  private val fhirPathEngine: FHIRPathEngine =
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    with(FhirContext.forCached(FhirVersionEnum.R4)) {
+      FHIRPathEngine(HapiWorkerContext(this, DefaultProfileValidationSupport(this)))
+    }
+}
+
+fun Type.valueOrCalculateValue(): Type? {
+  return if (this.hasExtension()) {
+    this.extension.firstOrNull { it.url == CQF_CALCULATED_EXPRESSION_URL }?.let {
+      val expression = (it.value as Expression).expression
+      ValidationUtil.fhirPathEngine.evaluate(this, expression).firstOrNull()?.let { it as Type }
```
Comment (shoaibmushtaq25) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt:40]: and change it from `ValidationUtil.fhirPathEngine.evaluate` to `fhirPathEngine.evaluate`

**Review by shoaibmushtaq25 — COMMENTED** [review]
@aurangzaibumer , I reviewed non-test classes. Left a few comments. Otherwise LGTM. cc: @f-odhiambo

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt (hunk 14 above)
Comment (aurangzaibumer) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt]: actually valueOrCalculatedValue is an extension that is out of scope of ValidationUtil that is why it was made internal so extension can access it. I have refactored ValidationUtil to use it as a file now.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt (hunk 15 above)
Comment (aurangzaibumer) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ValidationUtil.kt]: actually valueOrCalculatedValue is an extension that is out of scope of ValidationUtil that is why it was made internal so extension can access it. I have refactored ValidationUtil to use it as a file now. So now I can directly use fhirPathengine reference now.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt (hunk 6 above)
Comment (aurangzaibumer) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt:118]: I have used dynamic dateType (as an answer) to make it dynamic. this gives me an answerDate which is going to be dynamic with time. its a year before because this test is expected to throw a message. 

`val answerDate =
      DateType(
        SimpleDateFormat("yyyy-MM-dd")
          .parse(
            (DateTimeType.today()
              .apply {
                add(Calendar.YEAR, -1)
                add(Calendar.DAY_OF_MONTH, -1)
              }
              .valueAsString)
          )
      )`

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt (hunk 10 above)
Comment (aurangzaibumer) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:65]: I was trying to set a specific date on the date picker so that it can fail (out of the max range) for that I will have to use [PickerAction](https://github.com/vgrec/EspressoExamples/blob/master/app/src/androidTest/java/com/vgrec/espressoexamples/DateTimePickerTest.java) class that requires a separate library, I'm currently looking for any other alternative but this is what I found till yet.

Comment (aurangzaibumer) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:65]: maybe I can set min and max value before current's date. let me try some alternatives.

Comment (omarismail94) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:91]: What is this test doing? It seems to be just setting today's date right now and not validating that it is in the range set in the extension (-1 to 4). Can you add a test that shows this?

Can you also add a test that sets the minimum value higher than the maximum value to see what happens?

**Comment by aurangzaibumer** [comment]
> @aurangzaibumer please fix build errors (run spotlessApply)

done. thanks

**Review by omarismail94 — COMMENTED** [review]
Review complete. Comments left to address:

https://github.com/google/android-fhir/pull/1442/files#r951279669

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 16]
```diff
@@ -81,6 +92,140 @@ class QuestionnaireItemDatePickerViewHolderFactoryEspressoTest {
...
+    onView(withId(R.id.text_input_layout)).perform(clickIcon(true))
+    onView(allOf(withText("OK")))
+      .inRoot(isDialog())
+      .check(matches(isDisplayed()))
+      .perform(ViewActions.click())
+
+    val today = DateTimeType.today().valueAsString
+
+    assertThat(questionnaireItemView.answers.singleOrNull()?.valueDateType?.valueAsString)
+      .isEqualTo(today)
```
Comment (maimoonak) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:151]: add an assertion for validity as well

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 17]
```diff
@@ -81,6 +92,140 @@ class QuestionnaireItemDatePickerViewHolderFactoryEspressoTest {
...
+        QuestionnaireResponse.QuestionnaireResponseItemComponent(),
+        validationResult = NotValidated,
+        answersChangedCallback = { _, _, _ -> },
+      )
+
+    runOnUI { viewHolder.bind(questionnaireItemView) }
+
+    val exception =
+      Assert.assertThrows(IllegalArgumentException::class.java) {
+        onView(withId(R.id.text_input_layout)).perform(clickIcon(true))
```
Comment (maimoonak) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:220]: import Assert.assertThrows

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 18]
```diff
@@ -81,6 +92,140 @@ class QuestionnaireItemDatePickerViewHolderFactoryEspressoTest {
...
+      .perform(ViewActions.click())
+
+    val today = DateTimeType.today().valueAsString
+
+    assertThat(questionnaireItemView.answers.singleOrNull()?.valueDateType?.valueAsString)
+      .isEqualTo(today)
+  }
+
+  @Test
+  fun shouldNotSetDateInput_outsideRange() {
```
Comment (maimoonak) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:155]: add separate tests for value outside min and max range with error text assertion

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 19]
```diff
@@ -81,6 +92,140 @@ class QuestionnaireItemDatePickerViewHolderFactoryEspressoTest {
...
+
+    val exception =
+      Assert.assertThrows(IllegalArgumentException::class.java) {
+        onView(withId(R.id.text_input_layout)).perform(clickIcon(true))
+        onView(allOf(withText("OK")))
+          .inRoot(isDialog())
+          .check(matches(isDisplayed()))
+          .perform(ViewActions.click())
+      }
+    assertThat(exception.message).isEqualTo("currentPage cannot be after lastPage")
```
Comment (maimoonak) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:226]: error message does not match with the actual issue

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 20]
```diff
@@ -141,9 +144,32 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
...
           .setTitleText(R.string.select_date)
           .setSelection(selectedDate)
+          .setCalendarConstraints(getCalenderConstraint())
           .build()
       }
 
+      private fun getCalenderConstraint(): CalendarConstraints {
+        val min =
+          (MinValueConstraintValidator.getMinValue(questionnaireItemViewItem.questionnaireItem) as?
+              DateType)
```
Comment (maimoonak) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:154]: import MinValueConstraintValidator.getMinValue

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt [hunk 21]
```diff
@@ -91,4 +99,93 @@ class MinValueConstraintValidatorTest {
...
+    val validationResult =
+      MinValueConstraintValidator.validate(
+        questionnaireItem,
+        answers,
+        InstrumentationRegistry.getInstrumentation().context
+      )
+    val expectedDateRange =
+      (MinValueConstraintValidator.getMinValue(questionnaireItem) as? DateType)?.valueAsString
+    assertThat(validationResult.isValid).isFalse()
+    assertTrue(validationResult.message.equals("Minimum value allowed is:$expectedDateRange"))
```
Comment (maimoonak) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt:150]: use assertThat

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt [hunk 22]
```diff
@@ -89,17 +95,83 @@ class QuestionnaireItemDatePickerViewHolderFactoryEspressoTest {
...
+  }
+
+  @Test
+  fun shouldSetDateInput_withinRange() {
+    val questionnaireItemView =
+      QuestionnaireItemViewItem(
+        Questionnaire.QuestionnaireItemComponent().apply {
+          addExtension().apply {
+            url = "http://hl7.org/fhir/StructureDefinition/minValue"
+            val minDate = DateType(Date()).apply { add(Calendar.YEAR, -1) }
```
Comment (omarismail94) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:127]: What is the granularity of `minDate`?  I was able to set the date to August 1, even though that is outside of the year range. If today is Aug 25 2022, the `minDate` should be  Aug 25 2021.

I tried to set July 31 as well, but it correctly did not let me.

**Review by omarismail94 — COMMENTED** [review]
I've made one comment about a potential edge case

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt (hunk 22 above)
Comment (aurangzaibumer) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:127]: But when you select 1st August from Calendar as soon as you closes the date picker, the validation will trigger and show you the error on the edit text then minimum value date is 25 August. Date picker is not restricting the dates of that particular month but if user selects, it will be validated and the error will be shown.

Comment (aurangzaibumer) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:127]: I restricted min and max dates using CompositeDateValidator 
Here's the result, probably this is exactly what you were asking. Please correct me if I'm wrong. thanks

Considering 
my current date : 14th of Sept, 2022
minValue : "today() - 1 'months'" 
maxValue : "today() + 1 'months'"

https://user-images.githubusercontent.com/35099184/187844018-d2eed54d-1c9a-40f7-8bae-75c60d4b9cbe.mp4


cc : @f-odhiambo

**Review by maimoonak — APPROVED** [review]

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt (hunk 22 above)
Comment (aurangzaibumer) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:127]: can you please review @aditya-07

Comment (omarismail94) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:127]: Is the change to use `CompositeDateValidator` uploaded yet to GitHub? I do not see that class being used as part of this PR

Comment (aurangzaibumer) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactoryEspressoTest.kt:127]: Pushed the new changes, I was waiting for your feedback so that I could commit the changes if you agree to the new implementation. thanks Omar

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt [hunk 23]
```diff
@@ -158,14 +162,12 @@ internal object QuestionnaireItemDatePickerViewHolderFactory :
...
-          CalendarConstraints.Builder()
-            .apply {
-              min?.let { setStart(it) }
-              max?.let { setEnd(it) }
-            }
-            .build()
-        return constraintsBuilder
+        val listValidators = ArrayList<DateValidator>()
+        min?.let { listValidators.add(DateValidatorPointForward.from(it)) }
+        max?.let { listValidators.add(DateValidatorPointBackward.before(it)) }
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:167]: How granular are we allowing this? If someone set `min` to be granular up to seconds (i.e. 2022-09-12 15:05:44 UTC), should we allow it? Or should we just allow day as our most granular setting?

**Review by omarismail94 — COMMENTED** [review]
One more question about restricting date range

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt (hunk 23 above)
Comment (f-odhiambo) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:172]: Hey @omarismail94 I think we should limit this to just date (yyyy-mm-dd) and time (hh:mm). If all is ready we can close this Issue/PR and raise an enhancement ticket as a follow up   CC @aurangzaibumer @jingtang10

Comment (aurangzaibumer) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:172]: > How granular are we allowing this? If someone set `min` to be granular up to seconds (i.e. 2022-09-12 15:05:44 UTC), should we allow it? Or should we just allow day as our most granular setting?

I did try "valueDate": "2022-09-02T05:11:13" , the date picker ignored the time and used the date part as expected to restrict the dates.

Comment (f-odhiambo) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:172]: OK this is noted @aurangzaibumer seems this will only work with date , which is fine for this use case. @jingtang10 Kindly review and give feedback

**Review by omarismail94 — APPROVED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt (hunk 23 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemDatePickerViewHolderFactory.kt:172]: Thanks! Marking as resolved

**Comment by aurangzaibumer** [comment]
Hello @omarismail94 
I've tried updating the branch with master couple of times but PR gateways are still failing. 
Error Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 120000 ms
Can you please guide me on this. Thanks

**Comment by santosh-pingle** [comment]
For testing purpose updating example
``` {
  "resourceType": "Questionnaire",
  "item": [
    {
      "linkId": "1",
      "text": "When was your last menstrual period? (LMP)",
      "type": "date",
      "extension": [
        {
          "url": "http://hl7.org/fhir/StructureDefinition/maxValue",
          "valueDate": "2022-12-30"
        },
        {
          "url": "http://hl7.org/fhir/StructureDefinition/minValue",
          "valueDate": "2022-12-21"
        }
]
...
```
